### PR TITLE
Support named escapes (`\N{...}`) in string processing

### LINKED ISSUE #1468: Black produces code with invalid syntax by splitting `\n` in half
**Describe the bug**:
Black produces code with invalid syntax by splitting `\n` in half when reformatting string.

**To Reproduce**:
1. Put this code in a file:
```py
message = (
    f"1. Go to Google Developers Console and log in with your Google account.\n"
    "(https://console.developers.google.com/)\n"
    "2. You should be prompted to create a new project (name does not matter).\n"
    "3. Click on Enable APIs and Services at the top.\n"
    "4. In the list of APIs choose or search for YouTube Data API v3 and "
    "click on it. Choose Enable.\n"
    "5. Click on Credentials on the left navigation bar.\n"
    "6. Click on Create Credential at the top.\n"
    '7. At the top click the link for "API key".\n'
    "8. No application restrictions are needed. Click Create at the bottom.\n"
    "9. You now have a key to add to `[p]set api youtube api_key <your_api_key_here>`"
)
```
2. Run _Black_ on it with no arguments.
3. See error:
```
error: cannot format test.py: INTERNAL ERROR: Black produced invalid code: invalid syntax (<unknown>, line 2). Please report a bug on https://github.com/psf/black/issues.  This invalid output might be helpful: C:\Users\Jakub\AppData\Local\Temp\blk_6n5icmuc.log
```
Contents of the log file:
```diff
  File "c:\users\jakub\onedrive\dokumenty\red-discordbot\.venv\lib\site-packages\black\__init__.py", line 6037, in assert_equivalent
    dst_ast = parse_ast(dst)
  File "c:\users\jakub\onedrive\dokumenty\red-discordbot\.venv\lib\site-packages\black\__init__.py", line 5948, in parse_ast
    return ast27.parse(src)
  File "c:\users\jakub\onedrive\dokumenty\red-discordbot\.venv\lib\site-packages\typed_ast\ast27.py", line 50, in parse
    return _ast27.parse(source, filename, mode)
message = (
    f"1. Go to Google Developers Console and log in with your Google account.\n"
    f"(https://console.developers.google.com/)\"
    f"n2. You should be prompted to create a new project (name does not matter)."
    f"\n3. Click on Enable APIs and Services at the top"
    f".\n4. In the list of APIs choose or search for YouTube Data API v3 "
    f"and click on it. Choose Enab"
    f"le.\n5. Click on Credentials on the left navigation "
    f"bar.\n6. Click on Create Credential at the"
    f' top.\n7. At the top click the link for "API'
    f' key".\n8. No application restrictions are needed. Click Create at the '
    f"bottom.\n9. You now have a key to add to `[p]set api youtube api_key"
    f" <your_api_key_here>`"
)
```

**Expected behavior**
_Black_ shouldn't produce invalid syntax.

**Environment:**
- Version: master (commit 3461338)
- OS and Python version: Windows/Python 3.8.3

**Does this bug also happen on master?**:
It *only* happens on master, it doesn't happen on current release as there's no string reformatting in it yet. Online formatter doesn't error, but it does produce this output which is obviously invalid syntax:
```py
message = (
    f"1. Go to Google Developers Console and log in with your Google account.\n"
    f"(https://console.developers.google.com/)\"
    f"n2. You should be prompted to create a new project (name does not matter)."
    f"\n3. Click on Enable APIs and Services at the top"
    f".\n4. In the list of APIs choose or search for YouTube Data API v3 "
    f"and click on it. Choose Enab"
    f"le.\n5. Click on Credentials on the left navigation "
    f"bar.\n6. Click on Create Credential at the"
    f' top.\n7. At the top click the link for "API'
    f' key".\n8. No application restrictions are needed. Click Create at the '
    f"bottom.\n9. You now have a key to add to `[p]set api youtube api_key"
    f" <your_api_key_here>`"
)
```

**Additional context**
None

I'm sorry for the amount of issues but I tried to run Black's master on Cog-Creators/Red-DiscordBot's code base which resulted in quite a few of errors and also some undesired (by me) changes which I wanted to at least report even if they're not gonna be changed.

### PR DESCRIPTION
**Pull request by jack1142**
Fixes #1468 (or at least the issue that's in the comment there since they're not as closely related as I initially thought)

Not an awfully big amount of tests added here but I'm not sure if there's a need for more.

### CODE REVIEW

File: src/black/trans.py
Comment (felix-hilden) [inline src/black/trans.py:1266]: If I'm not mistaken, with escaped backslashes this would function incorrectly: `\\N{something}` would result in `count = 2` and still finding a slice. So how about something like:

```python
previous_was_backslash = False
for idx, c in it:
    if c == "\\":
        if not previous_was_backslash:
            previous_was_backslash = True
        else:
            previous_was_backslash = False    
        continue

    if not previous_was_backslash or c != "N":
        continue

    # Do the slice
```
Other potential names: `unescaped`, `previous_was_unescaped`, `previous_backslash_was_unescaped` (lol maybe not)

File: tests/data/long_strings.py
Comment (felix-hilden) [inline tests/data/long_strings.py:221]: Maybe it would be appropriate to test the escaping too, that would help in testing the backslash algorithm.

**Review by felix-hilden — COMMENTED** [review]
Hi, thanks for submitting! I'm not familiar with this part of the code base so take it with a grain of salt, but I left some thoughts below. I'd like for the more experienced maintainers to take a look too.

File: src/black/trans.py
Comment (jack1142) [reply src/black/trans.py:1353]: I mostly used this because I got suggested by the very similar function for f-strings above, but this one's definitely simpler so thanks! 😄

File: src/black/trans.py
Comment (jack1142) [reply src/black/trans.py:1266]: `backslash_count` was supposed to be used for escapes but looks I completely forgot to actually put such handling in...
I'll look into fixing this and adding more tests to cover it later today.

Comment (felix-hilden) [reply src/black/trans.py:1266]: Yeah I think testing for an odd number of backslashes works just as well if you were going for that!

Comment (felix-hilden) [reply src/black/trans.py:1266]: Actually now I see that my version wouldn't reset if some other character was after a backslash, so just keep yours I think 😅

Comment (jack1142) [reply src/black/trans.py:1269]: Escapes should work now, I decided to go for a tri-bool.

File: src/black/trans.py
Comment (felix-hilden) [inline src/black/trans.py:1263]: Is this redundant? Because `not None is True`, so this could be combined with below. I think. Then it would start flipping between `True` and `False`.

Comment (jack1142) [reply src/black/trans.py:1263]: It isn't because we want to do an early-continue *if* it's not a backslash.

Comment (felix-hilden) [reply src/black/trans.py:1263]: I get that you're signifying "no backslashes" with `None`, but `False` seems to do the same thing, since we don't care where the escaping happens. But at this point it's splitting hairs 😅 so feel free to ignore me.

Comment (jack1142) [reply src/black/trans.py:1263]: Although I guess I could change the order and simplify this a little bit:
```py
if c == "\\":
    backslash_tracking_state = not backslash_tracking_state
    continue
if backslash_tracking_state is None:
    continue
```

Comment (jack1142) [reply src/black/trans.py]: Okay, I see what you mean now 😄 Pushed a commit that should simplify this a fair bit.

**Review by felix-hilden — APPROVED** [review]
In my opinion this looks much better now, thanks a ton for being so speedy to modify!

**Comment by JelleZijlstra** [comment]
I wonder if this could be generalized into other string pieces that can't be split, so we don't need special-case logic for the different kinds. Instead, we could just generate a single list of unsplittable slices.

**Comment by felix-hilden** [comment]
Yep it should definitely be done!

**Comment by jack1142** [comment]
> I wonder if this could be generalized into other string pieces that can't be split, so we don't need special-case logic for the different kinds. Instead, we could just generate a single list of unsplittable slices.

I can look into this but before I do I would like to know if we care that some spans might overlap? Because I could either just make a list of slices from all functions that generate the slices (so `_get_nameescape_slices` and `fexpr_slices` at the current time) and just use that *or* I could go further - sort it and make a new list of non-overlapping spans.

**Comment by JelleZijlstra** [comment]
Overlap seems fine with how we're using these slices. Actually it may be more efficient to generate a set of illegal indices, since the current code looks quadratic to me. Which reminds me I should run some profiling for #2314.

**Comment by jack1142** [comment]
I went with the set idea as that indeed seems like a more performant way and isn't really hard to do either.

Sadly this causes the diff to be more complicated, if that's a problem I can split this into a separate PR that can be reviewed separately *after* this one is merged.

File: src/black/trans.py
Comment (JelleZijlstra) [inline src/black/trans.py:1300]:
```suggestion
                illegal_indices.update(range(begin, end + 1))
```

**Comment by felix-hilden** [comment]
If we wanted, I bet we could construct a single function that loops the string through once, cycling through N escape and f-string modes and generating the index ranges. But if this works, then it could be enough for this PR!

**Comment by ichard26** [comment]
Thank you so much for your contribution! This project is only possible by contributions like these :black_heart:.  You're awesome, @jack1142. Many thanks for pretty much beta testing the experimental string processing handling extremely early. I'm sure you're the reason why this feature will be a lot less buggy. Which is great since we want to push a release with it enabled by default soon.

**Pull request merged**
